### DISCUSSION TITLE
Hot reload and refresh problem

### DISCUSSION THREAD

**Discussion by ghost** [csells/go_router#201] [Q&A] [answered]
Hi,
In the example of "books" if I am located, for example, at "http://localhost:49325/#/authors" and I click on the refresh button of the browser, the page is changed to "http://localhost:49325/#/signin".
Also if I do a "hot reload" being on any page, the page is changed to "http://localhost:49325/#/signin":
This is a serious problem.
Thanks for your help.
Regards,
Jose

**Comment by csells** [comment, marked as answer]
Flutter web does not have stateful hot reload, so when you make a change, it does a hot refresh instead. When that happens, if you haven't persisted the user's login state, e.g. via [shared_preferences](https://pub.dev/packages/shared_preferences), then your user is not logged in and is redirected to the signin page. If you want to turn off redirection while debugging, you can do something like:

```dart
...
redirect: (state) {
  if (kDebugMode) return null;
  ...
}
..
```

Flutter web will use deep linking to keep you at the current page while you're doing hot refresh, so if you use this trick, you should stay on the same page as you make changes. Also, if you persist the user's credentials between sessions, either manually via something like shared_preferences or using Firebase Auth, which can automatically persist credentials between sessions depending on your app's settings, then that should work, too.

**Comment by jamolina2021** [comment]
@csells, Thanks for the explanation.

**Comment by osaxma** [comment]
I had the same issue before but the problem was caused by FirebaseAuth. 

In case you're using that, the problem was that `FirebaseAuth.instance.currentUser` would return `null` after `hot reload` or after refreshing the browser even if the user is signed in (issue filed [here](https://github.com/FirebaseExtended/flutterfire/issues/7404)). The `null` user caused my app to return the signing page since it assumes the user haven't signed in yet. 

To ensure that `currentUser` is not null for signed users,  the following in can be added to the `main()` function:
```dart
  await Firebase.initializeApp();
  if (kIsWeb) {
   // This will ensure that FirebaseAuth.instance.currentUser provides the correct user for the rest of the app. 
    await FirebaseAuth.instance.authStateChanges().first;
  }
```
